extensions/amp-story/1.0/progress-bar.js

```diff
@@ -119,10 +177,180 @@ export class ProgressBar {
       true /** callToInitialize */
     );
 
+    this.storeService_.subscribe(
+      StateProperty.RTL_STATE,
+      rtlState => {
+        this.onRtlStateUpdate_(rtlState);
+      },
+      true /** callToInitialize */
+    );
+
+    const ampdoc = Services.ampdocServiceFor(this.win_).getSingleDoc();
+    Services.viewportForDoc(ampdoc).onResize(
+      debounce(this.win_, () => this.onResize_(), 300)
+    );
+
+    this.barWidthPx_ = this.storyEl_
+      .querySelector('amp-story-page')
+      .getBoundingClientRect()./*OK*/ width;
+
+    this.vsync_.mutate(() => {
+      // Don't animate the progress bar in pageload.
+      this.timer_.delay(() => {
+        this.root_.classList.add('i-amphtml-animate-progress');
+      }, 60);
+    });
+
     this.isBuilt_ = true;
     return this.getRoot();
   }
 
+  /**
+   * Reacts to story replay.
+   * @private
+   */
+  replay_() {
+    this.progressBarIndex_ = 0;
+    this.render_();
+  }
+
+  /**
+   * Renders the segments by setting their corresponding width and translate.
+   * @private
+   */
+  render_() {
+    const segmentWidth = this.getSegmentWidth_();
+    let translateX =
+      -(this.progressBarIndex_ - this.getPrevEllipsisCount_()) *
+      (ELLIPSE_WIDTH_PX + SEGMENTS_MARGIN_PX);
+
+    for (let index = 0; index < this.segmentCount_; index++) {
```

Comment:
Can we wrap this entire loop in a `mutate` to avoid the overhead of calling `mutate` for each segment?

The big advantage is that you can get a promise out of `mutate`, that will resolve when the mutation has been executed.
This will be a useful heuristic to know when to add or not the `animate` class you have. You could think of adding a `shouldAnimate` optional parameter, and then toggle the class before/after the animation as needed.
It'd be great so we could remove the animation on pageload, resize, replay.

WDYT?